### DISCUSSION TITLE
Developing an iOS app with VS Code

### DISCUSSION THREAD

**Discussion by realh** [swiftlang/vscode-swift#360] [Q&A]
Hi,

I'd like to use VS Code to edit an iOS app I'm developing. I found [an article on Medium](https://medium.com/swlh/ios-development-on-vscode-27be37293fe1) but it refers to the old plugin that used to be included in the sourcekit-lsp repo. I've added `.vscode/settings.json` to my workspace, containing:

```json
{
    "sourcekit-lsp.serverArguments": [
        "-Xswiftc", "-sdk",
        "-Xswiftc", "/Applications/Xcode.app/Contents/Developer/Platforms/iPhoneSimulator.platform/Developer/SDKs/iPhoneSimulator.sdk",
        "-Xswiftc", "-target",
        "-Xswiftc", "arm64-apple-ios15.5-simulator"
    ]
}
```
and `Package.swift`:
```swift
// swift-tools-version:5.6
import PackageDescription
let packageName = "GroupCompanion"
let package = Package(
  name: packageName,
  platforms: [.iOS("15.5")],
  products: [
    .library(name: packageName, targets: [packageName])
  ],
  targets: [
    .target(
      name: packageName,
      path: packageName
    )
  ]
)
```
`swift package resolve` runs without errors, but I'm getting no support for code completion, linting, jump to definition etc, even for  types defined in my own code.

Can anyone help me get this to work?

**Comment by adam-fowler** [comment]
For sourcekit-lsp to work fully you need to compile your project. You are better setting up build arguments to compile your code. These are automatically passed onto sourcekit-lsp.

There's an interesting thread on build iOS projects using SPM here https://forums.swift.org/t/use-spm-to-build-ios-target/25436/

**Reply by adam-fowler** [reply]
These settings will allow you to build a project for iOS. You will need to set the SDK to whatever SDK you have installed.

```json
{
    "swift.buildArguments": [
        "--arch",
        "arm64",
        "-Xswiftc",
        "-sdk",
        "-Xswiftc",
        "/Applications/Xcode.app/Contents/Developer/Platforms/iPhoneOS.platform/Developer/SDKs/iPhoneOS15.5.sdk",
        "-Xswiftc",
        "-target",
        "-Xswiftc",
        "arm64-apple-ios14.0"
    ]
}
```

It appears XCTest does not work though. 
Once you have built the project, you should get symbol completion for iOS.

**Reply by adam-fowler** [reply]
Actually you should be able to reduce the settings to 
```json
{
    "swift.buildArguments": [
        "--arch",
        "arm64",
        "-Xswiftc",
        "-target",
        "-Xswiftc",
        "arm64-apple-ios15.0-simulator"
    ],
    "swift.SDK": "/Applications/Xcode.app/Contents/Developer/Platforms/iPhoneSimulator.platform/Developer/SDKs/iPhoneSimulator.sdk"
}
```

**Reply by realh** [reply]
Thanks. Looks like the main problem was that I was using the option name from the old plugin. It's mostly working with your corrections, but I need to add dependencies for grpc and Firebase. I added these as Swift Packages with XCode; is there a way to make the "dummy" Package.swift share those dependencies, or should I just add them to Package.swift and let SPM manage them independently of xcodeproj?

Also, I don't know how to ensure that I'm using a correct value for the option beginning "arm64". I just guessed the 15.5 part because that's the version that appears in the iOS Simulator's menu on my Mac, and on my iPhone. Is there a more reliable way to determine this option?

**Reply by 0xTim** [reply]
Pulling in dependencies defined in an Xcode project instead of a Swift Package manifest is probably where the extension will reach it's limit. Making that work would require us to start parsing the Xcode project file which is not something we'll do. You could try a dummy Package.swift file, no guarantees it would work though

**Reply by realh** [reply]
I'm already using a dummy Package.swift, because I thought that was necessary to make this work at all. I'll just duplicate the dependencies in that then.

**Reply by adam-fowler** [reply]
You can add the packages as dependencies to your dummy Package.swift. 

Regarding the 15.5 part of the simulator I think if you use the SDK without the version number it is a link to the latest SDK.

**Reply by adam-fowler** [reply]
You can also get the SDK path using `xcrun --sdk iphonesimulator --show-sdk-path`

**Reply by realh** [reply]
I added the dependencies manually and ran
`swift build --arch arm64 -Xswiftc -sdk -Xswiftc "$(xcrun --sdk iphonesimulator --show-sdk-path)" -Xswiftc -target -Xswiftc arm64-apple-ios15.5-simulator`
but ran into a couple of problems. The first is with FirebaseMessaging. I got this error:

`error: the library 'GroupCompanion' requires macos 10.10, but depends on the product 'FirebaseAuth' which requires macos 10.12; consider changing the library 'GroupCompanion' to require macos 10.12 or later, or the product 'FirebaseAuth' to require macos 10.10 or earlier.`

There doesn't seem to be a way to tell it I don't want to build for macos at all, and the only solution I could find was to change the `platforms` line in `Package.swift`:

`  platforms: [.iOS("15.5"), .macOS(.v10_12)],`

It fixed that problem, but I think this might be forcing a MacOS build, which could be the cause of the next error:

`/Users/`...`/UserAppDelegate.swift:26:9: error: cannot find 'Messaging' in scope`
`        Messaging.messaging().delegate = self`

`Messaging` is from Firebase. ISTR that FCM isn't supported on MacOS, so could it be that this part of the library just doesn't exist in MacOS builds? Or is there some other problem? These are the target dependencies:

```
      dependencies: [
        .product(name: "FirebaseAuth", package: "firebase-ios-sdk"),
        .product(name: "FirebaseMessaging", package: "firebase-ios-sdk"),
        .product(name: "GRPC", package: "grpc-swift"),
      ],
```

`FirebaseAuth` seems to work OK, I'm not getting any errors for the parts of my code which import that.

**Reply by realh** [reply]
This problem seems to have resolved itself now, but I've got a new problem, for which I'll start a new sub-thread.

**Comment by realh** [comment]
My new problem is that I'm now using Kotlin Multiplatform Mobile, and I want to include the framework it generates for my project's shared module in my dummy package. I found an article about including a `.binaryTarget` in a swift package, but this seems to only support a special `xcframework` format which is not the same as what KMM generates. The latter is a folder called `GCshared.framework`, containing a binary, `GCshared`, and an ObjC header, `Headers/GCshared.h`. I've tried adding the `Headers` path to -I on the command line, and even added -L and -l options for the binary, but that didn't work either.

Anyone know how I can get SPM to recognise this framework/header?

**Comment by adam-fowler** [comment]
Hi @realh I don't know a great deal about this kinda stuf. You are best asking in either the swift forums or the Swift Open Source Slack https://join.slack.com/t/swift-open-source/shared_invite/zt-1qos3yzfp-ZvU5GwTHB29pyAX9ah4q0A
